compiler/rustc_serialize/src/serialize.rs

```diff
@@ -65,10 +65,11 @@ pub trait Encoder {
         self.emit_u32(v as u32);
     }
 
+    // This is used for `&str` too.
     #[inline]
-    fn emit_str(&mut self, v: &str) {
+    fn emit_byte_str(&mut self, v: &[u8]) {
         self.emit_usize(v.len());
-        self.emit_raw_bytes(v.as_bytes());
+        self.emit_raw_bytes(v);
         self.emit_u8(STR_SENTINEL);
```

Comment:
`STR_SENTINEL` is now emitted for non UTF-8 strings, but `read_str` relies on it being an indication of UTF-8-ness to be somewhat safe.

Reply:
Does it? My mental model is that the emit/read stuff is type-directed. I.e. `read_str` will only be called on places where a utf8 string will be in the data because of the type structure.

Reply:
Okay, the serialization part is now no worse than it was previously.